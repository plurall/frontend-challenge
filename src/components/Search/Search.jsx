import { React, useState, useEffect } from 'react'
import { Link } from 'react-router-dom'
import { SubHeader } from 'components'
import { SomosClient } from 'utils'
import imageNotAvailable from '../../images/image-not-available.jpg'

export default function Search() {
  const [artistName, setArtistName] = useState('')
  const [artist, setArtist] = useState([])

  useEffect(() => {
    const searchArtist = async () => {
      if (artistName.length > 4) {
        const client = new SomosClient(artistName)
        const response = await client.getArtists()
        setArtist(response.artists.items)
      }
    }
    searchArtist()
  }, [artistName])

  return (
    <>
      <SubHeader
        breadcrumb={[{ text: 'Search' }]}
        heading="Desafio Front-end do Plurall"
      />
      <input
        type="text"
        placeholder="Digite o nome do artista"
        onChange={((e) => setArtistName(e.target.value))}
        className="input-artist-name"
      />
      <div
        className="above-the-card"
      >
        {artist.length !== 0 && artist.map(({ id, images, name }) => (
          <div
            key={id}
            className="card"
          >
            <Link
              to={`/artista/${id}`}
              className="href-name"
            >
              <img
                src={images.length !== 0 ? images[1].url : imageNotAvailable}
                alt="Artist"
                className="picture-album"
              />
              <p className="collection-name">{name}</p>
            </Link>
          </div>
      ))}
      </div>
    </>
  )
}

