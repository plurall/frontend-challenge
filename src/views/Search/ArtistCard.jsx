import React from 'react';

import { StyleArtistCard } from './styles';
import noAvatar from '../../assets/noAvatar.jpg'
import { useHistory } from 'react-router';


const ArtistCard = ({ artist }) => {
  const { images, id, name } = artist;

  const imageIndex = Math.floor(Math.random() * images.length);
  const avatarUrl = images.length ? images[imageIndex].url : null;

  const { push } = useHistory();

  const handleClick = e => push(`/artista/${id}`);

  return (
    <StyleArtistCard onClick={handleClick}>
      <span>{name}</span>
      <img src={avatarUrl || noAvatar} alt={`${name}-avatar`} />
    </StyleArtistCard>
  )
}

export default ArtistCard
