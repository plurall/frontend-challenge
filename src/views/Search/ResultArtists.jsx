import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';

import styles from './Search.module.css';
import ArtistImg from '../../assets/artist-avatar.png';

class ResultArtists extends React.Component {
  static propTypes = {
    artist: PropTypes.object,
  }

  render() {
    const {
      props: { artist },
    } = this;

    return (
      <>
        <Link to={`/artista/${artist.id}`}>
          <div className={styles.itemArtist}>
            <div className={styles.photoArtist}>
              {artist.images.length > 0 ? (
                <img src={artist.images[0].url} alt={artist.name} />
              ) : (
                <img src={ArtistImg} alt={artist.name} />
              )}
            </div>
            <p className={styles.titleArtist}>{artist.name}</p>
            <span className={styles.accessArtist}>Acessar</span>
          </div>
        </Link>
      </>
    );
  };
};

export default ResultArtists;
