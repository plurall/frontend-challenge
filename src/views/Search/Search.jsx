import React from 'react'
import { SearchBox } from '@plurall/elo'
import { Box, Layout } from 'components'

import SomosClient from 'utils/client'
import styles from './Search.module.css'

import { filterElement, filterNameToValue } from '../../utils'

class Search extends React.Component {
  state = {
    artists: [],
    data: '',
    list: [{ id: 0, value: '' }],
  }

  componentDidMount() {
    this.getArtistsData()
  }

  getArtistsData = async () => {
    const client = new SomosClient()
    const allArtists = await client.getArtists()
    const {
      data: { artists },
    } = allArtists

    this.setState(
      {
        artists,
      },
      () => console.log('ARTISTS', this.state.artists),
    )
  }

  handleChange = event => {
    if (event.length >= 4) {
      const data = filterElement(this.state.artists, event)
      const list = filterNameToValue(this.state.artists)
      this.setState(
        {
          data,
          list,
        },
        () => console.log(this.state.data),
      )
    }
  }

  render() {
    const { data, list } = this.state

    return (
      <Layout>
        <div className={styles.wrapper}>
          <SearchBox
            label="Digite o nome do artista"
            placeholder="Digite a sua busca ex: Daron"
            notFoundMessage="Nenhum resultado encontrado"
            items={[...list]}
            // onClick={item => console.log(item)}
            onChange={this.handleChange}
          />
          <Box data={data} />
        </div>
      </Layout>
    )
  }
}

export default Search
