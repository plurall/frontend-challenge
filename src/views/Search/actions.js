import { SomosClient } from '../../utils'

export function changeValue(e) {
  const params = { q: e.target.value, type: 'artist', market: 'us' }
  return dispatch => {
    dispatch({ type: 'CHANGE_VALUE', payload: e.target.value })
    if (e.target.value.length > 4) {
      SomosClient('search', params, res => {
        dispatch({
          type: 'HANDLE_SEARCH',
          payload: res.data.artists,
        })
      })
    }
  }
}

export function changeAlbums(id) {
  const params = { market: 'us', include_groups: 'single', limit: 10 }
  return dispatch => {
    SomosClient(`/artists/${id}/albums`, params, res => {
      dispatch({ type: 'HANDLE_ALBUMS', albums: res.data, id })
    })
  }
}
