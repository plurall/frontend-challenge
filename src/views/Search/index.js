export { default } from './Search'
export { changeValue, changeAlbums } from './actions'
export { default as searchReducer } from './reducer'
