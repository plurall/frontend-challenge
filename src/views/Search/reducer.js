const INITIAL_STATE = {}

export default function(state = INITIAL_STATE, action) {
  switch (action.type) {
    case 'HANDLE_ALBUMS':
      return {
        ...state,
        filter_artists: state.artists.items.filter(n => n.id === action.id),
        albums: action.albums.items,
        id: action.id,
      }
    case 'CHANGE_VALUE':
      return { change: action.payload }
    case 'HANDLE_SEARCH':
      return { artists: action.payload }
    default:
      return state
  }
}
