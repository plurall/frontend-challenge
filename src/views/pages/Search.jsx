import SearchFilter from 'components/SearchFilter/SearchFilter'
import React, { useState } from 'react'
import { SomosClient } from 'utils'

const client = new SomosClient()
const Search = () => {
  const [artists, setArtists] = useState([])
  const handleGetArtist = async name => {
    if (name.length > 4) {
      const data = await client.getArtists(name)
      setArtists(data.artists.items)
      console.log(data)
    }
  }
  return (
    <div>
      <SearchFilter onChange={handleGetArtist} />
      {artists?.length > 0 ? (
        <div>
          <ul>
            {artists?.map(artist => (
              <li>
                <img alt={artist.name} src={artist.images[0]?.url} />
                {artist.name}
              </li>
            ))}
          </ul>
        </div>) : (<div>Sem resultado</div>)}
    </div>
  )
}

export default Search
