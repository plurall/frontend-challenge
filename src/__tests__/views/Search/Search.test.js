import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import Search from 'views/Search';

describe('[Tela Busca]', () => {
  it('deve ser definido o componente', () => {

    expect(Search).toBeDefined();
  });

  it('deve equivaler ao snapshot', () => {
    const tree = shallow(
      <Search />
    );

    expect(tree).toMatchSnapshot();
  });

  it('deve chamar a função no onchange', () => {
    const wrapper = shallow(<Search />);

    wrapper.find('input.c-search__input').simulate('change', { target: {
      value: 'The Killers' }
    });

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("calls onBlur function when textField is blurred", () => {
    const wrapper = shallow(<Search />);
    const spy = jest.spyOn(wrapper.instance(), "handleSearch");

    wrapper.instance().forceUpdate();
    wrapper.find("input.c-search__input").simulate("change", spy);
    expect(spy).toBeCalled();
  });
});
