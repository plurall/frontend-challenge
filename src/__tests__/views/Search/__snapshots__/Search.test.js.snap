// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Tela Busca] deve chamar a função no onchange 1`] = `
<Layout>
  <SubHeader
    breadcrumb={
      Array [
        Object {
          "text": "Home / Busca",
        },
      ]
    }
    heading="Busca"
  />
  <div
    className="wrapper"
  >
    <input
      className="c-search__input"
      onChange={[Function]}
      placeholder="Busque pelo artista desejado..."
      type="search"
      value="The Killers"
    />
    <section
      className="c-search__result"
    />
    <Button
      href="/"
      text="Home"
    />
  </div>
</Layout>
`;

exports[`[Tela Busca] deve equivaler ao snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <SubHeader
          breadcrumb={
            Array [
              Object {
                "text": "Home / Busca",
              },
            ]
          }
          heading="Busca"
        />,
        <div
          className="wrapper"
        >
          <input
            className="c-search__input"
            onChange={[Function]}
            placeholder="Busque pelo artista desejado..."
            type="search"
            value=""
          />
          <section
            className="c-search__result"
          />
          <Button
            href="/"
            text="Home"
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "breadcrumb": Array [
            Object {
              "text": "Home / Busca",
            },
          ],
          "heading": "Busca",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="c-search__input"
              onChange={[Function]}
              placeholder="Busque pelo artista desejado..."
              type="search"
              value=""
            />,
            <section
              className="c-search__result"
            />,
            <Button
              href="/"
              text="Home"
            />,
          ],
          "className": "wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "c-search__input",
              "onChange": [Function],
              "placeholder": "Busque pelo artista desejado...",
              "type": "search",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": null,
              "className": "c-search__result",
            },
            "ref": null,
            "rendered": null,
            "type": "section",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "href": "/",
              "text": "Home",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <SubHeader
            breadcrumb={
              Array [
                Object {
                  "text": "Home / Busca",
                },
              ]
            }
            heading="Busca"
          />,
          <div
            className="wrapper"
          >
            <input
              className="c-search__input"
              onChange={[Function]}
              placeholder="Busque pelo artista desejado..."
              type="search"
              value=""
            />
            <section
              className="c-search__result"
            />
            <Button
              href="/"
              text="Home"
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "breadcrumb": Array [
              Object {
                "text": "Home / Busca",
              },
            ],
            "heading": "Busca",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="c-search__input"
                onChange={[Function]}
                placeholder="Busque pelo artista desejado..."
                type="search"
                value=""
              />,
              <section
                className="c-search__result"
              />,
              <Button
                href="/"
                text="Home"
              />,
            ],
            "className": "wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "c-search__input",
                "onChange": [Function],
                "placeholder": "Busque pelo artista desejado...",
                "type": "search",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": null,
                "className": "c-search__result",
              },
              "ref": null,
              "rendered": null,
              "type": "section",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "href": "/",
                "text": "Home",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
